src/App.tsx

```diff
@@ -36,35 +50,63 @@ function App(props: Props) {
       state.signer &&
       state.address &&
       roundId &&
-      (previousAddress === null || previousSigner === null)
+      (!previousAddress || !previousSigner)
     ) {
       const address = state.address;
       const message = `UMA Protocol one time key for round: ${roundId}`;
+      const hexMessage = web3.utils.toHex(message);
       const keyExists = signingKeys[address];
-      if (!keyExists || keyExists.roundMessage !== message) {
-        state.signer
-          .signMessage(message)
-          .then((msg) => {
-            const key = {} as {
-              publicKey: string;
-              privateKey: string;
-              roundMessage: string;
-            };
-
-            const privateKey = derivePrivateKey(msg);
-            const publicKey = recoverPublicKey(privateKey);
-            key.privateKey = privateKey;
-            key.publicKey = publicKey;
-            key.roundMessage = message;
-
-            setSigningKeys((prevKeys) => {
-              return { ...prevKeys, [address]: key };
+      let keysInStorage =
+        localStorage.getItem(SIGNING_KEYS_STORAGE_KEY) || ({} as StorageKeys);
+
+      let keyExistsInStorage = false;
+      if (typeof keysInStorage === "string") {
+        keysInStorage = JSON.parse(keysInStorage) as StorageKeys;
+
+        keyExistsInStorage = has(keysInStorage, `${hexMessage}.${address}`);
+      }
+
+      if (keyExistsInStorage) {
```

Comment:
it looks like storage always overrides whats in memory, so local storage acts as source of truth.  as long as theres no way to get out of sync between memory and local storage i think thats ok.  is that a safe assumption?

Reply:
The flow here is that you save new data into local storage and then update what's in state so react can re-render.

I think that's the only way this works. Because just referencing localStorage is insufficient for React to know to update state.

Reply:
ok cool that makes sense